PurchasesTests/Mocks/MockProductsManager.swift

```diff
@@ -9,22 +9,82 @@ import StoreKit
 
 class MockProductsManager: ProductsManager {
 
+    var invokedProductsFromOptimalStoreKitVersionWithIdentifiers = false
+    var invokedProductsFromOptimalStoreKitVersionWithIdentifiersCount = 0
+    var invokedProductsFromOptimalStoreKitVersionWithIdentifiersParameters: (identifiers: Set<String>, Void)?
+    var invokedProductsFromOptimalStoreKitVersionWithIdentifiersParametersList = [(identifiers: Set<String>, Void)]()
+    var stubbedProductsFromOptimalStoreKitVersionWithIdentifiersCompletionResult: (Set<ProductDetails>, Void)?
+
+    override func productsFromOptimalStoreKitVersion(withIdentifiers identifiers: Set<String>,
+        completion: @escaping (Set<ProductDetails>) -> Void) {
+        invokedProductsFromOptimalStoreKitVersionWithIdentifiers = true
+        invokedProductsFromOptimalStoreKitVersionWithIdentifiersCount += 1
+        invokedProductsFromOptimalStoreKitVersionWithIdentifiersParameters = (identifiers, ())
+        invokedProductsFromOptimalStoreKitVersionWithIdentifiersParametersList.append((identifiers, ()))
+        if let result = stubbedProductsFromOptimalStoreKitVersionWithIdentifiersCompletionResult {
+            completion(result.0)
+        } else {
+            let products: [SK1Product] = identifiers.map { (identifier) -> MockSK1Product in
+                let p = MockSK1Product(mockProductIdentifier: identifier)
+                p.mockSubscriptionGroupIdentifier = "1234567"
+                if #available(iOS 11.2, tvOS 11.2, macOS 10.13.2, *) {
+                    let mockDiscount = MockDiscount()
+                    mockDiscount.mockIdentifier = "discount_id"
+                    p.mockDiscount = mockDiscount
+                }
+                return p
+            }
+            let result = Set(products).map { SK1ProductDetails(sk1Product: $0) }
+
+            completion(Set(result))
+        }
+    }
+
+    var invokedProductsFromOptimalStoreKitVersion = false
+    var invokedProductsFromOptimalStoreKitVersionCount = 0
+    var invokedProductsFromOptimalStoreKitVersionParameters: (identifiers: Set<String>, Void)?
+    var invokedProductsFromOptimalStoreKitVersionParametersList = [(identifiers: Set<String>, Void)]()
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    override func productsFromOptimalStoreKitVersion(withIdentifiers identifiers: Set<String>) async -> Set<ProductDetails> {
+        invokedProductsFromOptimalStoreKitVersion = true
+        invokedProductsFromOptimalStoreKitVersionCount += 1
+        invokedProductsFromOptimalStoreKitVersionParameters = (identifiers, ())
+        invokedProductsFromOptimalStoreKitVersionParametersList.append((identifiers, ()))
+        let result = stubbedProductsFromOptimalStoreKitVersionWithIdentifiersCompletionResult?.0 ?? Set<ProductDetails>()
+        return result
+    }
```

Comment:
Doesn't really make sense to mock these two functions that are only available in iOS 15 right? We would never be able to mock the result since we can't create a stubbed set of `ProductDetails`

Reply:
I started a monologue here to talk about this https://github.com/RevenueCat/purchases-ios/pull/863/files?file-filters%5B%5D=.pbxproj&file-filters%5B%5D=.swift&hide-deleted-files=true#r713472002 if you want to get more context of what I am talking about

Reply:
we _can_ create stubbed productDetails, though: 
- for `SK1ProductDetails`, this can be a full mock
- for `SK2ProductDetails`, this can still be a mock, it just needs to have a valid `underlyingSK2Product`. but we could even mock out every method that accesses it

Reply:
How would you create a valid `underlyingSK2Product`?

Reply:
it'd be like this: 
https://github.com/RevenueCat/purchases-ios/pull/873/files#diff-a11a8960e5aca1406c6653742165482268006c78d678a511b2524bfa371bd0b0R87

- It has to be done through the `StoreKitUnitTests` target
- after you have the valid object, you can create a `MockSK2ProductDetails` and mock / spy all of its values if needed (I didn't for this particular test)